activesupport/lib/active_support/notifications/event_assertions.rb

```diff
@@ -0,0 +1,80 @@
+      #     end
+      #
+      def assert_instrumentation_event(name, payload = nil, &block)
+        events = capture_instrumentation_events(name, &block)
+        selected_events = events.select { |event| event.name == name }
+        assert_not_empty(selected_events, "No #{name} events were found.")
+
+        return if payload.nil?
+
+        event = selected_events.find { |event| event.payload == payload }
+        assert_not_nil(
+          event,
+          "No #{name} event with payload #{payload} was found. " \
+            "Did you mean one of these payloads: #{selected_events.map(&:payload).join("\n")}?",
+        )
+      end
+
+      # Assert the number of events emitted with the given +name+.
+      #
+      # You can assert the number of events emitted by passing a name, count,
+      # and block. While the block is executed, the number of matching events
+      # emitted will be counted. After the block is executed, the assertion
+      # will pass if the count matches. Otherwise, it will fail.
+      #
+      #     assert_instrumentation_events_count("post.submitted", 1) do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_events_count(name, count, &block)
+        events = capture_instrumentation_events(name, &block)
+        actual_count = events.select { |event| event.name == name }.count
+        assert_equal(count, actual_count, "Expected #{count} instead of #{actual_count} events for #{name}")
+      end
+
+      # Assert no events were emitted for the given +name+.
+      #
+      # You can assert no events were emitted by passing a name and block.
+      # While the block is executed, if no matching events are emitted,
+      # the assertion will pass. Otherwise, it will fail.
+      #
+      #     assert_no_instrumentation_events("post.submitted") do
+      #       post.destroy # => emits non-matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_no_instrumentation_events(name, &block)
+        events = capture_instrumentation_events(name, &block)
+        selected_events = events.select { |event| event.name == name }
+        assert_empty(selected_events, "Expected no events for #{name} but found #{selected_events.size}")
+      end
+
+      # Capture all events emitted for the given +name+.
+      #
+      # You can capture all events emitted filtered by a name by passing a name and block.
+      #
+      #     events = capture_instrumentation_events("post.submitted") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      def capture_instrumentation_events(name, &block)
+        events = []
+        ActiveSupport::Notifications.subscribed(->(e) { events << e }) { block.call }
+        events
+      end
```

Comment:
I'm sorry to have missed this before, but the `name` argument isn't being used in the current implementation.

The [ActiveSupport::Notifications#subscribed](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-subscribed) method accepts an optional `pattern` argument, which filters events by String or Regexp name.

Since `capture_instrumentation_events` can be used independent from the assertions, what do you think about:

* renaming `name` to `pattern` (to mirror the underlying method's argument name
* making `pattern` optional
* passing `pattern` to the `subscribed` call

```suggestion
      def capture_instrumentation_events(pattern = nil, &block)
        events = []
        ActiveSupport::Notifications.subscribed(->(e) { events << e }, pattern, &block)
        events
      end
```

That change has the potential to simplify the rest of the implementations, especially the ones that utilize `selected_events = events.select { |event| event.name == name }`.

Since the `capture_instrumentation_events` can handle the filtering, any object in the returned `events` array will be guaranteed to have the expected name.

Reply:
No worries, so I didn't notice either. Been some long days recently. :sweat_smile: 

I think renaming to `pattern` in `capture_instrumentation_events` makes sense, especially since this method is much closer to the actual event subscription and all, makes sense to shadow the underlying parameter. The built in filtering is also ideal cleanup wise.

Reply:
Just had the thought, suppose the other methods `name` param and docs should probably be updated to `pattern` as well so peeps know they can use such. Otherwise peeps likely think only string is supported.